server/resp/src/main/java/org/infinispan/server/resp/commands/hll/PFADD.java

```diff
@@ -0,0 +1,76 @@
+package org.infinispan.server.resp.commands.hll;
+
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.server.resp.Consumers;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespErrorUtil;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.commands.Resp3Command;
+import org.infinispan.server.resp.hll.HyperLogLog;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * The `<code>PFADD key [element [element ...]]</code>` command.
+ * <p>
+ * Adds an element to the HyperLogLog (HLL) structure. The command always verify if is possible to use the HLL structure
+ * with the given <code>key</code>. If the command is invoked without elements, the entry is created.
+ * </p>
+ *
+ * @see <a href="https://redis.io/commands/pfadd/">Redis documentation.</a>
+ * @since 15.0
+ * @author José Bolina
+ */
+public class PFADD extends RespCommand implements Resp3Command {
+
+   public PFADD() {
+      super(-2, 1, 1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      byte[] key = arguments.get(0);
+      AdvancedCache<byte[], Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
+      HyperLogLog hll = retrieveHLL(cache, key);
+
+      // The key is storing something which is not an HLL.
+      if (hll == null) {
+         RespErrorUtil.wrongType(handler.allocator());
+         return handler.myStage();
+      }
+
+      List<byte[]> elements = arguments.subList(1, arguments.size());
+      if (elements.isEmpty()) {
+         CompletionStage<Long> cs = cache.putIfAbsentAsync(key, hll).thenApply(v -> v == null ? 1L : 0L);
+         return handler.stageToReturn(cs, ctx, Consumers.LONG_BICONSUMER);
+      }
+
+      long before = hll.cardinality();
+      for (byte[] el : elements) {
+         hll.add(el);
+      }
+
+      return handler.stageToReturn(cache.putAsync(key, hll)
```

Comment:
This probably needs to be a `replace` call if the entry was there previously in case if a different node is also updating the HLL values.

Reply:
Updated utilizing the `compute` method instead. For the `replace`, we would need to keep the previous reference around.